gson/src/main/java/com/google/gson/stream/JsonReader.java

```diff
@@ -721,6 +722,7 @@ private int peekNumber() throws IOException {
     long value = 0; // Negative to accommodate Long.MIN_VALUE more easily.
     boolean negative = false;
     boolean fitsInLong = true;
+    int exponentDigitsCount = 0;
```

Comment:
What would you think of instead recording the index of the first exponent digit? Then you could still easily reject a number if it has too many exponent digits. Or you could parse the digits and compare against a limit, which would fix the leading-zero problem.

Reply:
> What would you think of instead recording the index of the first exponent digit?

I guess that would be possible, but it might then rely on the buffer not being refilled (and indices not changing) during parsing, which works because this is currently the case. Though from that perspective tracking the count seems a bit more reliable to me.

Is your concern performance (I assume not)? Or whether tracking the index might lead to cleaner or easier to understand code?

> Or you could parse the digits and compare against a limit, which would fix the leading-zero problem.

I mainly omitted the leading 0s check for simplicity (but nonetheless added a comment and test to document the behavior). Fixing it might also be possible by slightly adjusting this method to explicitly check for `0`. For example:
```diff
 ...
 } else if (last == NUMBER_CHAR_EXP_E || last == NUMBER_CHAR_EXP_SIGN) {
   last = NUMBER_CHAR_EXP_DIGIT;
+  if (c != '0') {
     exponentDigitsCount++;
+  }
 } else if (last == NUMBER_CHAR_EXP_DIGIT) {
+  if (exponentDigitsCount > 0 || c != '0') {
     exponentDigitsCount++;
 
     // Similar to the scale limit enforced by NumberLimits.parseBigDecimal(String)
     // This also considers leading 0s (e.g. '1e000001'), but probably not worth the effort ignoring them
     if (exponentDigitsCount > 4) {
       throw new MalformedJsonException("Too many number exponent digits" + locationString());
     }
+  }
}
```

I just wasn't sure if that many leading 0s is really a common use case and worth supporting.
Should I solve this though with the diff shown above (and some comments)?

Reply:
The parsing logic is fairly complicated, but it seems to me that the way the `i` variable works is quite simple, starting from 0 and increasing. Even if the buffer is refilled and `pos` changes, `i` is still a valid offset. So I think saving its value and using it at the end is reasonably robust. And I feel the resulting code would be slightly simpler. It also seems very slightly better to check for a specific maximum exponent rather than a number of exponent digits.

I still have the question of whether we need to change `JsonReader` at all, if we're also checking for large exponents in `TypeAdapters`.

Reply:
> It also seems very slightly better to check for a specific maximum exponent rather than a number of exponent digits.

Ok, I will give it a try.

> I still have the question of whether we need to change `JsonReader` at all, if we're also checking for large exponents in `TypeAdapters`.

Changing `JsonReader` is mainly for the cases where users directly obtain the number from the `JsonReader` using `nextString()` and then parse the number themselves. Do you think this should not be done? In that case is it ok if I change the documentation of `nextString()` to tell the users to validate the number string, if necessary?

Reply:
Obsolete, I have reverted the changes to `JsonReader`, see https://github.com/google/gson/pull/2510#discussion_r1358951745